File: Project Netwerken/src/NodeJS/global/model/packet/PacketPart.js
const varint = require('../../utils/varint/varint');

class PacketPart {
    constructor(options) {
        if (options.buffer) {
            this.id = varint.decode(options.buffer);
            this.currentPart = varint.decode(options.buffer);
            this.total = varint.decode(options.buffer);
            let offset = options.buffer.position();
            let fullarray = options.buffer.array;
            this.commandPart = varint.decode(fullarray.slice(offset, fullarray.length));
        } else {
            this.id = options.id;
            this.currentPart = options.currentPart;
            this.total = options.total;
            this.commandPart = options.commandPart;
        }
    }

/*    // OLD CONSTRUCTOR
    constructor(buffer) {
        this.id = varint.decode(buffer);
        this.currentPart = varint.decode(buffer);
        this.total = varint.decode(buffer);
        let offset = buffer.position();
        let fullarray = buffer.array;
        this.commandPart = varint.decode(fullarray.slice(offset, fullarray.length));
    }

    // OLD CONSTRUCTOR
	constructor(id, currentPart, total, commandPart) {
		this.id = id;
		this.currentPart = currentPart;
		this.total = total;
		this.commandPart = commandPart;
	}*/

    write(buffer){
        buffer.write(varint.encode(this.id));
        buffer.write(varint.encode(this.currentPart));
        buffer.write(varint.encode(this.total));
        buffer.write(this.commandPart);
    }
}

File: Project Netwerken/src/AndroidApp/app/src/main/java/com/example/loraapp/global/model/packet/PacketPart.java
package com.example.loraapp.global.model.packet;

import com.example.loraapp.global.utils.VarInt;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.Arrays;

/**
 * PacketPart to send the packet in parts when needed.
 */
public class PacketPart {
    private long id = -1;
    private int currentPart;
    private int total;
    private byte[] commandPart;

    public PacketPart(byte[] bytes) {
        ByteBuffer buffer = ByteBuffer.wrap(bytes);
        this.id = VarInt.readUnsignedVarLong(buffer);
        this.currentPart = VarInt.readUnsignedVarInt(buffer);
        this.total = VarInt.readUnsignedVarInt(buffer);
        byte[] fullArray = buffer.array();
        int offset = buffer.position();
        this.commandPart = Arrays.copyOfRange(fullArray,offset, fullArray.length);
    }
    /**
     * @param id serie id of parts
     * @param currentPart current commandPart number of total
     * @param total number of parts
     * @param commandPart commandPart of encryptedCommand
     */
    public PacketPart(long id, int currentPart, int total, byte[] commandPart) {
        this.id = id;
        this.currentPart = currentPart;
        this.total = total;
        this.commandPart = commandPart;
    }

    public long getId() {
        return id;
    }

    /**
     * @return current commandPart number of total
     */
    public int getCurrentPart() {
        return currentPart;
    }

    /**
     * @return number of parts
     */
    public int getTotal() {
        return total;
    }

    /**
     * @return commandPart of encryptedCommand
     */
    public byte[] getCommandPart() {
        return commandPart;
    }

    public byte[] write() throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        VarInt.writeUnsignedVarLong(id, out);
        out.write(VarInt.writeUnsignedVarInt(currentPart));
        out.write(VarInt.writeUnsignedVarInt(total));
        out.write(commandPart);
        return out.toByteArray();
    }
}
